🐛 [Bug]: cache middleware: runtime error: index out of range [0] with length 0

### Bug Description

Use cache middleware to cache home page "/".
Use CacheInvalidator to skip cache, got error: runtime error: index out of range [0] with length 0.

### How to Reproduce

1.use cache middleware to cache home page "/".
2.do not visit "/" in browser, so there will not be any cache of it.
3.visit "/login" page to login, then CacheInvalidator would return true now.
4.visit "/".
5.got error.

### Expected Behavior

When restart the program, not every page is cached, when user logged in visit no cache page, it should be open normal.

### Fiber Version

v3.0.0-beta.3

### Code Snippet (optional)

```go
var WebCache_Home = cache.New(cache.Config{
	CacheInvalidator: auth.ValidateLogin,
	Expiration: 10 * time.Minute,
	KeyGenerator: func(c fiber.Ctx) string {
		cacheKey := "homeLogout"
		if auth.ValidateLogin(c){
			cacheKey = "homeLogin"
		}
		return cacheKey
	},
	StoreResponseHeaders: true,
	MaxBytes:             1 * 1024 * 1024,
})

app.Use("/:p<maxLen(0)>?", cache.WebCache_Home)
app.Get("/:p<maxLen(0)>?", user.HomeHandler)
```


### Checklist:

- [X] I agree to follow Fiber's [Code of Conduct](https://github.com/gofiber/fiber/blob/master/.github/CODE_OF_CONDUCT.md).
- [X] I have checked for existing issues that describe my problem prior to opening this one.
- [X] I understand that improperly formatted bug reports may be closed without explanation.

> @canbusio We need an example to replicate this, from your snippet we can't tell what is `auth.ValidateLogin`. is it using anything from the context to valdiate?
> 
> Cam yo share the runtime error logs

Hi~

I tried again with auth.ValidateLogin codes bellow:

func ValidateLogin(c fiber.Ctx) bool {
    return true
}

And this "index out of range [0] with length 0" error can be reproduced.

By the way, when I remove "CacheInvalidator: auth.ValidateLogin", it works fine.


The full run time error log is like this:

panic: runtime error: index out of range [0] with length 0

goroutine 55 [running]:
github.com/gofiber/fiber/v3/middleware/cache.(*indexedHeap).remove(0xc0002e2db0?, 0xe?)
        /home//go/pkg/mod/github.com/gofiber/fiber/v3@v3.0.0-beta.3/middleware/cache/heap.go:86 +0xb8
github.com/gofiber/fiber/v3/middleware/cache.New.func4({0xb6fb78, 0xc00023a008})
        /home/user/go/pkg/mod/github.com/gofiber/fiber/v3@v3.0.0-beta.3/middleware/cache/cache.go:129 +0x348
github.com/gofiber/fiber/v3.(*App).next(0xc000250008, 0xc00023a008)
        /home/user/go/pkg/mod/github.com/gofiber/fiber/v3@v3.0.0-beta.3/router.go:188 +0x1c2
github.com/gofiber/fiber/v3.(*DefaultCtx).Next(0xc0002100c0?)
        /home/user/go/pkg/mod/github.com/gofiber/fiber/v3@v3.0.0-beta.3/ctx.go:1036 +0x79
github.com/gofiber/fiber/v3/middleware/limiter.FixedWindow.New.func1({0xb6fb78, 0xc00023a008})
        /home/user/go/pkg/mod/github.com/gofiber/fiber/v3@v3.0.0-beta.3/middleware/limiter/limiter_fixed.go:83 +0x1a7
github.com/gofiber/fiber/v3.(*App).next(0xc000250008, 0xc00023a008)
        /home/user/go/pkg/mod/github.com/gofiber/fiber/v3@v3.0.0-beta.3/router.go:188 +0x1c2
github.com/gofiber/fiber/v3.(*DefaultCtx).Next(0xc0000ee2a8?)
        /home/user/go/pkg/mod/github.com/gofiber/fiber/v3@v3.0.0-beta.3/ctx.go:1036 +0x79
flawer_leaf/internal/route.Setup.func1({0xb6fb78, 0xc00023a008})
        /home/user/Go/FlawerNext/Flawer_Leaf/internal/route/route.go:21 +0x24
github.com/gofiber/fiber/v3.(*App).next(0xc000250008, 0xc00023a008)
        /home/user/go/pkg/mod/github.com/gofiber/fiber/v3@v3.0.0-beta.3/router.go:188 +0x1c2
github.com/gofiber/fiber/v3.(*App).requestHandler(0xc000250008, 0x4ac7af?)
        /home/user/go/pkg/mod/github.com/gofiber/fiber/v3@v3.0.0-beta.3/router.go:236 +0x2f8
github.com/valyala/fasthttp.(*Server).serveConn(0xc0000f6248, {0xb69318, 0xc0002077c8})
        /home/user/go/pkg/mod/github.com/valyala/fasthttp@v1.55.0/server.go:2379 +0xe70
github.com/valyala/fasthttp.(*workerPool).workerFunc(0xc000144140, 0xc0002920e0)
        /home/user/go/pkg/mod/github.com/valyala/fasthttp@v1.55.0/workerpool.go:224 +0xa4
github.com/valyala/fasthttp.(*workerPool).getCh.func1()
        /home/user/go/pkg/mod/github.com/valyala/fasthttp@v1.55.0/workerpool.go:196 +0x32
created by github.com/valyala/fasthttp.(*workerPool).getCh in goroutine 83
        /home/user/go/pkg/mod/github.com/valyala/fasthttp@v1.55.0/workerpool.go:195 +0x190
exit status 2

--- a/middleware/cache/cache.go
+++ b/middleware/cache/cache.go
@@ -117,46 +117,49 @@ func New(config ...Config) fiber.Handler {
 		// Get timestamp
 		ts := atomic.LoadUint64(&timestamp)
 
-		// Invalidate cache if requested
-		if cfg.CacheInvalidator != nil && cfg.CacheInvalidator(c) && e != nil {
-			e.exp = ts - 1
-		}
-
-		// Check if entry is expired
-		if e.exp != 0 && ts >= e.exp {
-			deleteKey(key)
-			if cfg.MaxBytes > 0 {
-				_, size := heap.remove(e.heapidx)
-				storedBytes -= size
-			}
-		} else if e.exp != 0 && !hasRequestDirective(c, noCache) {
-			// Separate body value to avoid msgp serialization
-			// We can store raw bytes with Storage 👍
-			if cfg.Storage != nil {
-				e.body = manager.getRaw(key + "_body")
-			}
-			// Set response headers from cache
-			c.Response().SetBodyRaw(e.body)
-			c.Response().SetStatusCode(e.status)
-			c.Response().Header.SetContentTypeBytes(e.ctype)
-			if len(e.cencoding) > 0 {
-				c.Response().Header.SetBytesV(fiber.HeaderContentEncoding, e.cencoding)
-			}
-			for k, v := range e.headers {
-				c.Response().Header.SetBytesV(k, v)
-			}
-			// Set Cache-Control header if enabled
-			if cfg.CacheControl {
-				maxAge := strconv.FormatUint(e.exp-ts, 10)
-				c.Set(fiber.HeaderCacheControl, "public, max-age="+maxAge)
+		// Cache Entry not found
+		if e != nil {
+			// Invalidate cache if requested
+			if cfg.CacheInvalidator != nil && cfg.CacheInvalidator(c) {
+				e.exp = ts - 1
 			}
 
-			c.Set(cfg.CacheHeader, cacheHit)
+			// Check if entry is expired
+			if e.exp != 0 && ts >= e.exp {
+				deleteKey(key)
+				if cfg.MaxBytes > 0 {
+					_, size := heap.remove(e.heapidx)
+					storedBytes -= size
+				}
+			} else if e.exp != 0 && !hasRequestDirective(c, noCache) {
+				// Separate body value to avoid msgp serialization
+				// We can store raw bytes with Storage 👍
+				if cfg.Storage != nil {
+					e.body = manager.getRaw(key + "_body")
+				}
+				// Set response headers from cache
+				c.Response().SetBodyRaw(e.body)
+				c.Response().SetStatusCode(e.status)
+				c.Response().Header.SetContentTypeBytes(e.ctype)
+				if len(e.cencoding) > 0 {
+					c.Response().Header.SetBytesV(fiber.HeaderContentEncoding, e.cencoding)
+				}
+				for k, v := range e.headers {
+					c.Response().Header.SetBytesV(k, v)
+				}
+				// Set Cache-Control header if enabled
+				if cfg.CacheControl {
+					maxAge := strconv.FormatUint(e.exp-ts, 10)
+					c.Set(fiber.HeaderCacheControl, "public, max-age="+maxAge)
+				}
 
-			mux.Unlock()
+				c.Set(cfg.CacheHeader, cacheHit)
 
-			// Return response
-			return nil
+				mux.Unlock()
+
+				// Return response
+				return nil
+			}
 		}
 
 		// make sure we're not blocking concurrent requests - do unlock
@@ -193,6 +196,7 @@ func New(config ...Config) fiber.Handler {
 			}
 		}
 
+		e = manager.acquire()
 		// Cache response
 		e.body = utils.CopyBytes(c.Response().Body())
 		e.status = c.Response().StatusCode()

--- a/middleware/cache/cache_test.go
+++ b/middleware/cache/cache_test.go
@@ -47,9 +47,10 @@ func Test_Cache_Expired(t *testing.T) {
 	t.Parallel()
 	app := fiber.New()
 	app.Use(New(Config{Expiration: 2 * time.Second}))
-
+	count := 0
 	app.Get("/", func(c fiber.Ctx) error {
-		return c.SendString(strconv.FormatInt(time.Now().UnixNano(), 10))
+		count++
+		return c.SendString(strconv.Itoa(count))
 	})
 
 	resp, err := app.Test(httptest.NewRequest(fiber.MethodGet, "/", nil))
@@ -86,9 +87,10 @@ func Test_Cache(t *testing.T) {
 	app := fiber.New()
 	app.Use(New())
 
+	count := 0
 	app.Get("/", func(c fiber.Ctx) error {
-		now := strconv.FormatInt(time.Now().UnixNano(), 10)
-		return c.SendString(now)
+		count++
+		return c.SendString(strconv.Itoa(count))
 	})
 
 	req := httptest.NewRequest(fiber.MethodGet, "/", nil)
@@ -305,9 +307,10 @@ func Test_Cache_Invalid_Expiration(t *testing.T) {
 	cache := New(Config{Expiration: 0 * time.Second})
 	app.Use(cache)
 
+	count := 0
 	app.Get("/", func(c fiber.Ctx) error {
-		now := strconv.FormatInt(time.Now().UnixNano(), 10)
-		return c.SendString(now)
+		count++
+		return c.SendString(strconv.Itoa(count))
 	})
 
 	req := httptest.NewRequest(fiber.MethodGet, "/", nil)
@@ -414,8 +417,10 @@ func Test_Cache_NothingToCache(t *testing.T) {
 
 	app.Use(New(Config{Expiration: -(time.Second * 1)}))
 
+	count := 0
 	app.Get("/", func(c fiber.Ctx) error {
-		return c.SendString(time.Now().String())
+		count++
+		return c.SendString(strconv.Itoa(count))
 	})
 
 	resp, err := app.Test(httptest.NewRequest(fiber.MethodGet, "/", nil))
@@ -447,12 +452,16 @@ func Test_Cache_CustomNext(t *testing.T) {
 		CacheControl: true,
 	}))
 
+	count := 0
 	app.Get("/", func(c fiber.Ctx) error {
-		return c.SendString(time.Now().String())
+		count++
+		return c.SendString(strconv.Itoa(count))
 	})
 
+	errorCount := 0
 	app.Get("/error", func(c fiber.Ctx) error {
-		return c.Status(fiber.StatusInternalServerError).SendString(time.Now().String())
+		errorCount++
+		return c.Status(fiber.StatusInternalServerError).SendString(strconv.Itoa(errorCount))
 	})
 
 	resp, err := app.Test(httptest.NewRequest(fiber.MethodGet, "/", nil))
@@ -508,9 +517,11 @@ func Test_CustomExpiration(t *testing.T) {
 		return time.Second * time.Duration(newCacheTime)
 	}}))
 
+	count := 0
 	app.Get("/", func(c fiber.Ctx) error {
+		count++
 		c.Response().Header.Add("Cache-Time", "1")
-		return c.SendString(strconv.FormatInt(time.Now().UnixNano(), 10))
+		return c.SendString(strconv.Itoa(count))
 	})
 
 	resp, err := app.Test(httptest.NewRequest(fiber.MethodGet, "/", nil))
@@ -588,8 +599,11 @@ func Test_CacheHeader(t *testing.T) {
 		return c.SendString(fiber.Query[string](c, "cache"))
 	})
 
+	count := 0
 	app.Get("/error", func(c fiber.Ctx) error {
-		return c.Status(fiber.StatusInternalServerError).SendString(time.Now().String())
+		count++
+		c.Response().Header.Add("Cache-Time", "1")
+		return c.Status(fiber.StatusInternalServerError).SendString(strconv.Itoa(count))
 	})
 
 	resp, err := app.Test(httptest.NewRequest(fiber.MethodGet, "/", nil))
@@ -615,10 +629,13 @@ func Test_Cache_WithHead(t *testing.T) {
 	app := fiber.New()
 	app.Use(New())
 
+	count := 0
 	handler := func(c fiber.Ctx) error {
-		now := strconv.FormatInt(time.Now().UnixNano(), 10)
-		return c.SendString(now)
+		count++
+		c.Response().Header.Add("Cache-Time", "1")
+		return c.SendString(strconv.Itoa(count))
 	}
+
 	app.Route("/").Get(handler).Head(handler)
 
 	req := httptest.NewRequest(fiber.MethodHead, "/", nil)
@@ -708,8 +725,10 @@ func Test_CacheInvalidation(t *testing.T) {
 		},
 	}))
 
+	count := 0
 	app.Get("/", func(c fiber.Ctx) error {
-		return c.SendString(time.Now().String())
+		count++
+		return c.SendString(strconv.Itoa(count))
 	})
 
 	resp, err := app.Test(httptest.NewRequest(fiber.MethodGet, "/", nil))
@@ -729,6 +748,93 @@ func Test_CacheInvalidation(t *testing.T) {
 	bodyInvalidate, err := io.ReadAll(respInvalidate.Body)
 	require.NoError(t, err)
 	require.NotEqual(t, body, bodyInvalidate)
+}
+
+func Test_CacheInvalidation_noCacheEntry(t *testing.T) {
+	t.Parallel()
+	t.Run("Cache Invalidator should not be called if no cache entry exist ", func(t *testing.T) {
+		t.Parallel()
+		app := fiber.New()
+		cacheInvalidatorExecuted := false
+		app.Use(New(Config{
+			CacheControl: true,
+			CacheInvalidator: func(c fiber.Ctx) bool {
+				cacheInvalidatorExecuted = true
+				return fiber.Query[bool](c, "invalidate")
+			},
+			MaxBytes: 10 * 1024 * 1024,
+		}))
+		_, err := app.Test(httptest.NewRequest(fiber.MethodGet, "/?invalidate=true", nil))
+		require.NoError(t, err)
+		require.False(t, cacheInvalidatorExecuted)
+	})
+}
+
+func Test_CacheInvalidation_removeFromHeap(t *testing.T) {
+	t.Parallel()
+	t.Run("Invalidate and remove from the heap", func(t *testing.T) {
+		t.Parallel()
+		app := fiber.New()
+		app.Use(New(Config{
+			CacheControl: true,
+			CacheInvalidator: func(c fiber.Ctx) bool {
+				return fiber.Query[bool](c, "invalidate")
+			},
+			MaxBytes: 10 * 1024 * 1024,
+		}))
+
+		count := 0
+		app.Get("/", func(c fiber.Ctx) error {
+			count++
+			return c.SendString(strconv.Itoa(count))
+		})
+
+		resp, err := app.Test(httptest.NewRequest(fiber.MethodGet, "/", nil))
+		require.NoError(t, err)
+		body, err := io.ReadAll(resp.Body)
+		require.NoError(t, err)
+
+		respCached, err := app.Test(httptest.NewRequest(fiber.MethodGet, "/", nil))
+		require.NoError(t, err)
+		bodyCached, err := io.ReadAll(respCached.Body)
+		require.NoError(t, err)
+		require.True(t, bytes.Equal(body, bodyCached))
+		require.NotEmpty(t, respCached.Header.Get(fiber.HeaderCacheControl))
+
+		respInvalidate, err := app.Test(httptest.NewRequest(fiber.MethodGet, "/?invalidate=true", nil))
+		require.NoError(t, err)
+		bodyInvalidate, err := io.ReadAll(respInvalidate.Body)
+		require.NoError(t, err)
+		require.NotEqual(t, body, bodyInvalidate)
+	})
+}
+
+func Test_CacheStorage_CustomHeaders(t *testing.T) {
+	t.Parallel()
+	app := fiber.New()
+	app.Use(New(Config{
+		CacheControl: true,
+		Storage:      memory.New(),
+		MaxBytes:     10 * 1024 * 1024,
+	}))
+
+	app.Get("/", func(c fiber.Ctx) error {
+		c.Response().Header.Set("Content-Type", "text/xml")
+		c.Response().Header.Set("Content-Encoding", "utf8")
+		return c.Send([]byte("<xml><value>Test</value></xml>"))
+	})
+
+	resp, err := app.Test(httptest.NewRequest(fiber.MethodGet, "/", nil))
+	require.NoError(t, err)
+	body, err := io.ReadAll(resp.Body)
+	require.NoError(t, err)
+
+	respCached, err := app.Test(httptest.NewRequest(fiber.MethodGet, "/", nil))
+	require.NoError(t, err)
+	bodyCached, err := io.ReadAll(respCached.Body)
+	require.NoError(t, err)
+	require.True(t, bytes.Equal(body, bodyCached))
+	require.NotEmpty(t, respCached.Header.Get(fiber.HeaderCacheControl))
 }
 
 // Because time points are updated once every X milliseconds, entries in tests can often have

--- a/middleware/cache/heap.go
+++ b/middleware/cache/heap.go
@@ -15,7 +15,7 @@ type heapEntry struct {
 // elements in constant time. It does so by handing out special indices
 // and tracking entry movement.
 //
-// indexdedHeap is used for quickly finding entries with the lowest
+// indexedHeap is used for quickly finding entries with the lowest
 // expiration timestamp and deleting arbitrary entries.
 type indexedHeap struct {
 	// Slice the heap is built on

--- a/middleware/cache/manager.go
+++ b/middleware/cache/manager.go
@@ -83,8 +83,7 @@ func (m *manager) get(key string) *item {
 		return it
 	}
 	if it, _ = m.memory.Get(key).(*item); it == nil { //nolint:errcheck // We store nothing else in the pool
-		it = m.acquire()
-		return it
+		return nil
 	}
 	return it
 }

--- a/middleware/cache/manager_test.go
+++ b/middleware/cache/manager_test.go
@@ -1,0 +1,26 @@
+package cache
+
+import (
+	"testing"
+	"time"
+
+	"github.com/gofiber/utils/v2"
+	"github.com/stretchr/testify/assert"
+)
+
+func Test_manager_get(t *testing.T) {
+	t.Parallel()
+	cacheManager := newManager(nil)
+	t.Run("Item not found in cache", func(t *testing.T) {
+		t.Parallel()
+		assert.Nil(t, cacheManager.get(utils.UUID()))
+	})
+	t.Run("Item found in cache", func(t *testing.T) {
+		t.Parallel()
+		id := utils.UUID()
+		cacheItem := cacheManager.acquire()
+		cacheItem.body = []byte("test-body")
+		cacheManager.set(id, cacheItem, 10*time.Second)
+		assert.NotNil(t, cacheManager.get(id))
+	})
+}

